xdsl/transforms/printf_to_putchar.py

```diff
@@ -0,0 +1,243 @@
+            digit = arith.RemUI(i_1, ten)
+            ascii_offset = arith.Constant.from_int_and_width(ord("0"), i32)
+            char = arith.Addi(digit, ascii_offset)
+            char_i8 = arith.TruncIOp(char, i8)
+            PrintCharOp((char_i8,))
+            scf.Yield.get()
+        return for_loop
+
+    # Beginning of inline_itoa_func declaration
+    """
+    Python code for this implementation:
+    ```
+    if(integer = 0):
+        print("0")
+    else:
+        absolute_value = print_minus_if_negative(integer)
+        digits = get_number_of_digits(absolute_value)
+        print_digits(digits, absolute_value)
+    ```
+    """
+    inline_itoa_func = func.FuncOp("inline_itoa", ((i32,), ()))
+    with ImplicitBuilder(inline_itoa_func.body) as (integer,):
+        zero = arith.Constant.from_int_and_width(0, i32)
+        is_zero_block = Block()
+        is_not_zero_block = Block()
+        is_zero = arith.Cmpi(integer, zero, "eq")
+        scf.If.get(
+            is_zero,
+            [],
+            [is_zero_block],
+            [is_not_zero_block],
+        )
+        # If the value is zero, just print one zero
+        with ImplicitBuilder(is_zero_block):
+            PrintCharOp.from_constant_char("0")
+            scf.Yield.get()
+        # Otherwise continue with itoa
+        with ImplicitBuilder(is_not_zero_block):
+            # Print minus sign if negative and return absolute value
+            absolute_value = print_minus_if_negative(integer)
+            # Get amount  of digits to be printed
+            digits = get_number_of_digits(absolute_value)
+            print_digits(digits, absolute_value)
+            # Yield from is_not_zero_block
+            scf.Yield.get()
+        # Return from itoa function
+        func.Return()
+
+    return inline_itoa_func
+
+
+if __name__ == "__main__":
+    printer = Printer()
+    printer.print(get_inline_itoa())
+
+
+class LowerPrintCharToPutcharPass(ModulePass):
+    name = "lower-printchar-to-putchar"
+
+    # lower to func.call
+    def apply(self, ctx: MLContext, op: ModuleOp) -> None:
+        print_int_walker = PatternRewriteWalker(
+            GreedyRewritePatternApplier(
+                [
+                    ConvertPrintIntToItoa(),
+                ]
+            ),
+            apply_recursively=True,
+        )
+        print_int_walker.rewrite_module(op)
+        # This has to happen first, since inline itoa
+        # contains some references to printf.print_char
+        op.body.block.add_ops([get_inline_itoa()])
+        # Add function implementation for inline_itoa
+        # TODO Don't insert this if the operation is not there!
+        print_char_walker = PatternRewriteWalker(
+            GreedyRewritePatternApplier([LowerPrintCharToPutchar()]),
+            apply_recursively=True,
+        )
+        print_char_walker.rewrite_module(op)
+        # Add external putchar reference
+        # TODO Don't insert this if the operation is not there!
```

Comment:
Would be good to add this functionality as part of this PR. it should be pretty easy with the symbol table functionality, or just a manual scan of the module op body.

Reply:
Could you give a reference/example to the symbol table functionality? I don't know it. Thanks!

Reply:
It might not work here, but in theory can be something like in the `get_op_for_symbol` function in interpreter.py

``` python


    def get_op_for_symbol(self, symbol: str) -> Operation:
        if symbol in self.symbol_table:
            return self.symbol_table[symbol]
        else:
            raise InterpretationError(f'Could not find symbol "{symbol}"')
```

Reply:
If it doesn't work I'd be interested :)

Reply:
Can you doublecheck now